paper-linear-progress.html

```diff
@@ -0,0 +1,897 @@
+         * representing the percentage complete of buffering. Note that the value will be clamped
+         * if it is less than/greater than 0/100, respectively.
+         */
+        bufferValue: {
+          type: Number,
+          observer: '_bufferValueChanged',
+          value: 100
+        },
+
+        /**
+         * When set, progress bars animate across the element, expressing an operation which takes
+         * an unknown amount of time.
+         */
+        indeterminate: {
+          type: Boolean,
+          observer: '_indeterminateChanged',
+          // We give a default value so that correct ARIA attributes are applied when element is
+          // initialized.
+          value: false
+        },
+
+        /**
+         * When set, will reverse all animations, transforms, and transitions occuring within the
+         * element. Useful when working with RTL languages, as well as distinguishing between phases
+         * of an operation.
+         */
+        reversed: {
+          type: Boolean,
+          observer: '_reversedChanged'
+        },
+
+        /**
+         * When set, will display buffering dots which appear to animate across the element. Useful
+         * when displaying progress for operations involving media, during which content may need to
+         * be buffered.
+         */
+        buffering: {
+          type: Boolean,
+          observer: '_bufferingChanged'
+        },
+
+        /**
+         * When set, the linear progress element will be hidden from view. Note that this is
+         * _different_ from `hidden`, as it does not set `display: none;` but rather just changes
+         * the element's opacity.
+         */
+        closed: {
+          type: Boolean,
+          observer: '_closedChanged'
+        },
+
+        /**
+         * Animation configuration for NeonAnimationRunnerBehavior.
+         */
+        animationConfig: {
+          value: function() {
+            // NOTE(traviskaufman): This is initialized within the attached() method, once we know
+            // the height of the progress indicator.
+            return {};
+          }
+        },
+
+        __needsBufferingAnimationOnAttached: {
+          // HACK: We want to have buffering play an animation when it's set, but we can't
+          // animate unless we first have the computed height of the element, which we can only
+          // retrieve once the element's attached. Therefore, we set up a queue of animations that
+          // need to be run, and if we encounter any animations. Note that this can be fixed once
+          // https://github.com/Polymer/polymer/issues/2399 is resolved. Then, we can do all
+          // animations within CSS.
+          type: Boolean,
+          value: false
+        }
+      },
+
+      hostAttributes: {
+        role: 'progressbar'
+      },
+
+      ready: function() {
+        this.toggleClass('isIE', /MSIE/.test(navigator.userAgent));
```

Comment:
Thanks again for the thoughtful consideration here.

> It requires the creation and performance of an animating element, separately from handling the animation for the actual element itself.

And my guess is that this imposes asynchrony, which complicates things.

> Whether or not the element will be in a state reflective of the initial keyframe block when the event is fired would have to be deterministic.

This is easy enough to discover through experimentation.

> There would have to be a way of computing the element's transform within the animation, since the underlying CSS value would not change. This blog post outlines a way to do this, but due to a getComputedStyle call within a requestAnimationFrame block it would induce massive performance overhead.

There are a lot of unanswered questions here that would probably be answered nicely by a little experimentation. My questions:

 - Is it possible to measure the transform and get an actionable value back?
 - Does a useful measurement *have* to be within a `requestAnimationFrame`, or is it possible to perform synchronously? 
 - What unit are we using for the straw man quantity of "massive"? In other words, what is the relative performance cost of a one-time feature test compared to, say, instantiating hundreds of instances of some common Polymer element?

> it seems like UA-sniffing is an appropriate approach.

Matching User Agent strings is perilous and unhygienic, and should always be considered a last resort.

I would like us to experiment with feature detection and discover the boundaries of the problem. I'm happy to contribute by experimenting if you would find that helpful.

Ultimately, it's possible that any attempt to feature test will be met with more trouble than it is worth. In that case, it would be nice if we could discuss and consider alternative designs that will allow us to avoid UA detection.

WDYT?